optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -186,20 +186,109 @@
+     * values resulting from applying the group key mapping function on all items in the original stream. None of the
+     * items in the new stream will {@link Objects#equals(Object, Object)} any other.
+     *
+     * @param groupKeyMapping never null, mapping function to convert each element in the stream to a different element.
+     * @param <GroupKey_> the type of a match fact in the destination {@link UniConstraintStream}.
+     * @return never null.
+     */
+    <GroupKey_> UniConstraintStream<GroupKey_> groupBy(Function<A, GroupKey_> groupKeyMapping);
+
+    /**
+     * Convert the {@link UniConstraintStream} to a {@link BiConstraintStream}, consisting of:
+     *
+     * <ul>
+     *     <li>on the left, the set of values resulting from applying the group key mapping function on all items in
+     *     the original stream. None of the items in the new stream will {@link Objects#equals(Object, Object)} any
+     *     other.</li>
+     *     <li>on the right, the value of applying a given {@link UniConstraintCollector} on all the items in the
+     *     original {@link UniConstraintStream} with a matching group key.</li>
+     * </ul>
+     *
+     * @param groupKeyMapping never null, mapping function to convert each element in the stream to a different element.
+     * @param <GroupKey_> the type of the left match fact in the destination {@link BiConstraintStream}.
+     * @param <ResultContainer_> the mutable accumulation type (often hidden as an implementation detail)
+     * @param <Result_> the type of the right match fact in the destination {@link BiConstraintStream}
+     * @return never null.
+     */
     <GroupKey_, ResultContainer_, Result_> BiConstraintStream<GroupKey_, Result_> groupBy(
             Function<A, GroupKey_> groupKeyMapping,
             UniConstraintCollector<A, ResultContainer_, Result_> collector);
 
+    /**
+     * Convert the {@link UniConstraintStream} to a {@link BiConstraintStream}, consisting of:
+     *
+     * <ul>
+     *     <li>on the left, the set of values resulting from applying the "A" group key mapping function on all items in
+     *     the original stream. None of the items in the new stream will {@link Objects#equals(Object, Object)} any
+     *     other.</li>
+     *     <li>on the right, the set of values resulting from applying the "B" group key mapping function on all items
+     *     in the original stream, which already passed through the "A" group key mapping. None of the items in the new
+     *     stream will {@link Objects#equals(Object, Object)} any other.</li>
+     * </ul>
+     *
+     * Here's an example. Assuming:
+     *
+     * <ul>
+     *     <li>The original {@link UniConstraintStream} contains items "X1", "X2", "Y1"</li>.
+     *     <li>The "A" group key mapping function returns "X" for "X1" and "X2", "Y" for "Y1".</li>
+     *     <li>The "B" group key mapping function returns "1" for "X1" and "Y1", "2" for "A2".</li>
+     * </ul>
```

Comment:
So we remove this example?

If we keep it, let's use the syntax we agreed: {} for tuples, [] for elements